Dtls handshake serialization test

AndrzejKurek opened the pull request:
This PR adds a handshake serialization test on top of the previous PR with DTLS tests. Please review only the last commit (80a94ec).

piotr-now commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,6 +3192,42 @@ void handshake( const char *cipher, int version, int pk_alg,
+                                          &context_buf_len ) == 0 );
+
+    TEST_ASSERT( mbedtls_ssl_context_load( &(server.ssl), context_buf,
+                                           context_buf_len ) == 0 );
```
It would not be better if we do for example `mbedtls_ssl_free` between  `mbedtls_ssl_context_save` and `mbedtls_ssl_context_load` to make sure serialization works? Or use `mbedtls_ssl_context_load` on another object?

AndrzejKurek replied:
There are two layers of serialization used in `ssl_server2` serialization. This is the first one, and the second one  adds additional `mbedtls_ssl_free`, `mbedtls_ssl_init` followed by a bunch of setup stuff. I can add this too if we want to.

Patater replied:
Could we memset the buffer to all zeros in-between saving and loading at least? I'd like to have more confidence that the saved context doesn't have any old data in it that gets reused.

piotr-now commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,6 +3192,42 @@ void handshake( const char *cipher, int version, int pk_alg,
+
+    TEST_ASSERT( mbedtls_ssl_context_load( &(server.ssl), context_buf,
+                                           context_buf_len ) == 0 );
+    mbedtls_free( context_buf );
```
If one of two upper TEST_ASSERT will fail this `mbedtls_free` would not be handled

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,6 +3192,42 @@ void handshake( const char *cipher, int version, int pk_alg,
+        goto exit;
+
+    memset( cli_buf, 'X', MSGLEN );
+    memset( srv_buf, 'Y', MSGLEN );
```
That's the point. I'm setting it to a known value, so that if the buffer is not fully overwritten, I would catch that.

piotr-now replied:
Ok, I got it :).

piotr-now commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,9 +3194,48 @@ void handshake( const char *cipher, int version, int pk_alg,
+    if( dtls != 0 && serialize != 0 )
+        mbedtls_free( context_buf );
+#endif
 }
```
It could be simpler `if( context_buf != NULL )`, but it LGTM

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,9 +3194,48 @@ void handshake( const char *cipher, int version, int pk_alg,
+
+    TEST_ASSERT( mbedtls_ssl_context_save( &(server.ssl), context_buf,
+                                          context_buf_len,
+                                          &context_buf_len ) == 0 );
```
Style: alignment is off here.

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,9 +3194,48 @@ void handshake( const char *cipher, int version, int pk_alg,
     mbedtls_endpoint_free( &server, dtls != 0 ? &server_context : NULL );
+#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION)
+    if( dtls != 0 && serialize != 0 )
+        mbedtls_free( context_buf );
```
Style: Would prefer explicit `{}`

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,9 +3194,48 @@ void handshake( const char *cipher, int version, int pk_alg,
 
+#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION)
+    if( dtls == 0 || serialize == 0 )
+        goto exit;
```
Style: Would prefer explicit `{}`

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3184,9 +3194,48 @@ void handshake( const char *cipher, int version, int pk_alg,
     TEST_ASSERT( server.ssl.state == MBEDTLS_SSL_HANDSHAKE_OVER );
 
+#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION)
+    if( dtls == 0 || serialize == 0 )
```
Add a C code comment to explain why you want to skip this test.

AndrzejKurek commented:
I've added braces, a comment mentioned by Jaeden, fixed indentation and changed the serialization to be of mode 2, as described here: https://github.com/ARMmbed/mbedtls/blob/a08e699afcd5d4f420d6d91bdd05fce2f4db1330/programs/ssl/ssl_server2.c#L4114

AndrzejKurek commented:
Changed only the commit description.
